src/reducers/versions.tsx

```diff
@@ -1301,6 +1320,40 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
         },
       };
     }
+    case getType(actions.setLastVersionId): {
+      return {
+        ...state,
+        lastVersionId: action.payload.versionId,
+      };
+    }
+    case getType(actions.restoreExpandedPaths): {
+      const { versionId, lastVersionId } = action.payload;
+      const version = state.versionInfo[versionId];
+      const lastVersion = state.versionInfo[lastVersionId];
+
+      if (!version) {
+        throw new Error(`Version missing for versionId: ${versionId}`);
+      }
+      if (!lastVersion) {
+        throw new Error(`Version missing for versionId: ${lastVersionId}`);
+      }
+
+      const expandedPaths = lastVersion.expandedPaths.filter((path) =>
+        version.entries.find((e) => e.path === path),
+      );
+
+      return {
+        ...state,
+        lastVersionId: undefined,
+        versionInfo: {
+          ...state.versionInfo,
+          [versionId]: {
+            ...version,
+            expandedPaths,
```

Comment:
This makes sense but instead of introducing `setLastVersionId` and `restoreExpandedPaths`, could you maybe accomplish the same thing by using `currentVersionId`?

Here is a thought:

```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 6252178..678f8fc 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1176,12 +1176,26 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
     case getType(actions.loadVersionInfo): {
       const { version } = action.payload;
 
+      const newVersion = createInternalVersion(version);
+      let { expandedPaths, selectedPath } = newVersion;
+
+      const currentVersion = selectCurrentVersionInfo(state);
+      if (currentVersion && currentVersion.id === version.id) {
+        expandedPaths = currentVersion.expandedPaths;
+        selectedPath = currentVersion.selectedPath;
+      }
+
       return {
         ...state,
         versionInfo: {
           ...state.versionInfo,
-          [version.id]: createInternalVersion(version),
+          [version.id]: {
+            ...newVersion,
+            expandedPaths,
+            selectedPath,
+          },
         },
+        // TODO: maybe also set currentVersionId here?
       };
     }
     case getType(actions.abortFetchVersion): {

```

I added a note about possibly setting `currentVersionId` but I'm not sure if that's needed or not. Maybe not.

I haven't tried out this patch, it was just a thought inspired by your approach 😁

Reply:
Great. There is no need to introducing a new state variable if we use `currentVersionId`. But It looks like I still need one new action `restoreExpandedPath`. Putting this logic into `loadVersionInfo` does not work, because `loadVersionInfo` is not always dispatched when switching to a new version id (not when it is already loaded).